### DISCUSSION TITLE
Make HTTPBearer Return 401 over 403

### DISCUSSION THREAD

**Discussion by JulianHysi** [fastapi/fastapi#12384] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# auth.py
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials

security = HTTPBearer()


def verify_token(
    credentials: HTTPAuthorizationCredentials = Security(security),
) -> dict[str, str]:
    # code that interacts with my IDP service to validate the token
    # the issue persists no matter what the body of this function is
    # therefore I'm not including it


# routes/healthcheck.py
from fastapi import APIRouter, Depends

from src.auth import verify_token
from src.schemas import HealthCheckSchema

router = APIRouter(tags=["Healthcheck"])


@router.get(
    "/health-check",
    response_model=HealthCheckSchema,
    dependencies=[Depends(verify_token)],
)
async def healthcheck() -> HealthCheckSchema:
    return HealthCheckSchema(status=True)
```

### Description

The problem is that by default, If I don't pass a token at all to the request, the HTTPBearer class returns a 403 automatically.

Since I'm not passing the token, I'd expect a 401, not a 403. Or at least make this easily configurable via a param. As it is, my only option seems to be to override the class and return the 401 myself.

### Operating System

Linux

### FastAPI Version

0.112.2

### Pydantic Version

2.8.2

### Python Version

3.12.5

**Comment by sinisaos** [comment, marked as answer]
@JulianHysi Based on [#2026](https://github.com/fastapi/fastapi/issues/2026) and [#10177](https://github.com/fastapi/fastapi/issues/10177) nothing will change. I also think [401](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/401) is the correct status code for an invalid (or unset) token. Other frameworks like [Litestar](https://github.com/litestar-org/litestar/blob/e7dfc7aba2377dabc0ca53d9a1e41795481f5606/litestar/exceptions/http_exceptions.py#L98-L101) or [Blacksheep](https://github.com/Neoteroi/BlackSheep/blob/026897ff2796c2f2074cb368ff9c59a262dc28ad/blacksheep/server/authorization/__init__.py#L104-L114) have this correct. I don't think you have a better option than (as you said) to override `HTTPBearer` and return a `401` yourself, or to use `auto_error=False` and return `401` yourself in the endpoint.

**Reply by JulianHysi** [reply]
That's a bit of a bummer though.

**Reply by YuriiMotov** [reply]
This was fixed in 0.122.0 ([link](https://fastapi.tiangolo.com/release-notes/#01220)).
Now `401` status code is used in case of missing credentials

**Comment by Kfir-G** [comment]
Hey @JulianHysi,

The default behavior of FastAPI's `HTTPBearer` returning a 403 Forbidden when no token is provided can be a bit confusing since 403 usually means the server understood the request but refuses to authorize it. In cases where the token is simply missing, a 401 Unauthorized response is definitely more fitting.

While FastAPI doesn’t currently offer a built-in way to change this behavior in `HTTPBearer`, you can create a custom dependency to handle the authentication check yourself. Here’s a quick example of how you can do that:

### Custom Dependency Example

```python
from fastapi import APIRouter, Depends, HTTPException, status
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials

security = HTTPBearer()
router = APIRouter(tags=["Healthcheck"])

def verify_token(credentials: HTTPAuthorizationCredentials = Depends(security)):
    if credentials is None:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Not authenticated",
        )
    # Add your token validation logic here.
    # If the token is invalid, you can raise a 403 Forbidden error if necessary.

@router.get("/health-check", response_model=HealthCheckSchema, dependencies=[Depends(verify_token)])
async def healthcheck() -> HealthCheckSchema:
    return HealthCheckSchema(status=True)
```

### What This Does:
1. **Custom Dependency**: The `verify_token` function checks if the credentials are provided. If they’re not, it raises a `401 Unauthorized` error.
2. **Route Integration**: You can use this custom dependency in your route, replacing the default behavior.

### Benefits:
- With this approach, if a token isn’t provided, you’ll get a `401 Unauthorized` response, making it clearer to the user that authentication is required.

For more information, you can check out these: https://github.com/fastapi/fastapi/issues/2026

**Reply by JulianHysi** [reply]
Yup, I know how to make this work via custom code, but I'd expect it to work as it should out of the box.
